ktor-server/ktor-server-netty/jvm/src/io/ktor/server/netty/cio/NettyHttpResponsePipeline.kt

```diff
@@ -0,0 +1,345 @@
+     */
+    private fun processOrderedCall(call: NettyApplicationCall, block: () -> Unit) {
+        call.response.responseReady.addListener responseFlag@{ responseFlagResult ->
+            call.previousCallFinished.addListener waitPreviousCall@{ previousCallResult ->
+                if (!previousCallResult.isSuccess) {
+                    processCallFailed(call, previousCallResult.cause())
+                    return@waitPreviousCall
+                }
+                if (!responseFlagResult.isSuccess) {
+                    processCallFailed(call, responseFlagResult.cause())
+                    return@waitPreviousCall
+                }
+                block.invoke()
+            }
+        }
+    }
+
+    private fun processCallFailed(call: NettyApplicationCall, actualException: Throwable) {
+        val t = when {
+            actualException is IOException && actualException !is ChannelIOException ->
+                ChannelWriteException(exception = actualException)
+            else -> actualException
+        }
+
+        call.response.responseChannel.cancel(t)
+        call.responseWriteJob.cancel()
+        call.response.cancel()
+        call.dispose()
+        call.callFinished.setFailure(t)
+    }
+
+    private fun processUpgradeResponse(call: NettyApplicationCall, responseMessage: Any): ChannelFuture {
+        val future = context.write(responseMessage)
+        call.upgrade(context)
+        call.isByteBufferContent = true
+
+        context.flush()
+        isDataNotFlushed.compareAndSet(expect = true, update = false)
+        return future
+    }
+
+    private fun finishCall(
+        call: NettyApplicationCall,
+        lastMessage: Any?,
+        lastFuture: ChannelFuture
+    ) {
+        val prepareForClose = call.isContextCloseRequired() &&
+            (!call.request.keepAlive || call.response.isUpgradeResponse())
+
+        val lastMessageFuture = if (lastMessage != null) {
+            val future = context.write(lastMessage)
+            isDataNotFlushed.compareAndSet(expect = false, update = true)
+            future
+        } else {
+            null
+        }
+
+        httpHandler.activeRequests.decrementAndGet()
+        call.callFinished.setSuccess()
+
+        lastMessageFuture?.addListener {
+            if (prepareForClose) {
+                close(lastFuture)
+                return@addListener
+            }
+        }
+        if (prepareForClose) {
+            close(lastFuture)
+            return
+        }
+        scheduleFlush()
+    }
+
+    fun close(lastFuture: ChannelFuture) {
+        context.flush()
+        isDataNotFlushed.compareAndSet(expect = true, update = false)
+        lastFuture.addListener {
+            context.close()
+        }
+    }
+
+    private fun scheduleFlush() {
+        context.executor().execute {
+            flushIfNeeded()
+        }
+    }
+
+    private fun processCall(call: NettyApplicationCall) {
```

Comment:
please consider renaming and commenting: this method is getting the request message and decide how it should be handled